components/script/dom/htmlimageelement.rs

```diff
@@ -1032,3 +1053,228 @@ fn image_dimension_setter(element: &Element, attr: LocalName, value: u32) {
     let value = AttrValue::Dimension(value.to_string(), dim);
     element.set_attribute(&attr, value);
 }
+
+#[derive(Debug)]
+#[derive(PartialEq)]
+pub struct ImageSource {
+    pub url: String,
+    pub descriptor: Descriptor,
+}
+#[derive(PartialEq)]
+#[derive(Debug)]
+pub struct Descriptor {
+    pub wid: Option<u32>,
+    pub den: Option<f64>,
+}
+
+pub fn collect_sequence_characters<F>(s: &str, predicate: F) -> (&str, &str)
+        where F: Fn(&char) -> bool
+        {
+            for (i, ch) in s.chars().enumerate() {
+                if !predicate(&ch) {
+                    return (&s[0..i], &s[i..])
+                }
+            }
+        return (s, "");
+        }
+
+pub fn parse_a_srcset_attribute(input: String) -> Vec<ImageSource> {
+    let mut start = 0;
+    let mut candidates: Vec<ImageSource> = Vec::new();
+    loop {
+        let position = &input[start..];
+
+        let(spaces, position) = collect_sequence_characters(position, |c| *c ==',' || char::is_whitespace(*c));
+        let x = spaces.find(',');
+        match x {
+            Some(val) => println!("Parse Error"),
+            None => println!("No commas\n"),
+        }
+
+        // add the counts of spaces that we collect to advance the start index
+        let space_len = spaces.char_indices().count();
+        start += space_len;
+
+        //Returns and breaks out of the loop
+        if position.is_empty() {
+            return candidates;
+        }
+        let(url, spaces) = collect_sequence_characters(position, |c| !char::is_whitespace(*c));
+
+        // add the counts of urls that we parse to advance the start index
+        start += url.char_indices().count();
```

Comment:
Why should we use `char_indices` instead of `chars` here? (and below?). We should prefer the latter, unless we're using the indices during the iteration.